### ISSUE TITLE
Gff3 upload - Tripal4 

### ISSUE THREAD

**Issue by alslonik** [tripal/tripal#1437]
Hi, 
I am trying to upload a gff3 file with tripal4 gff file loader and get the following warnings and an error:

```
Running 'Chado GFF3 File Loader' importer
 [notice] Running 'Chado GFF3 File Loader' importer
NOTE: Loading of this file is performed using a database transaction. If it fails or is terminated prematurely then all insertions and updates are rolled back and will not be found in the database
 [notice] NOTE: Loading of this file is performed using a database transaction. If it fails or is terminated prematurely then all insertions and updates are rolled back and will not be found in the database
 [warning] Undefined array key "add_only" GFF3Importer.php:540
 [warning] Undefined array key "update" GFF3Importer.php:541
Opening /var/www/html/Drupal/web/sites/default/files/tripal/users/1/wonderful_29K_genes.gff
 [notice] Opening /var/www/html/Drupal/web/sites/default/files/tripal/users/1/wonderful_29K_genes.gff
Opening temporary cache file: /tmp/TripalGFF3Import_BPDhlt
 [notice] Opening temporary cache file: /tmp/TripalGFF3Import_BPDhlt
Step  1 of 27: Caching GFF3 file...                             
 [notice] Step  1 of 27: Caching GFF3 file...                             
 [warning] Undefined array key "uniquename" GFF3Importer.php:1206
Removing temporary cache file: /tmp/TripalGFF3Import_BPDhlt
 [notice] Removing temporary cache file: /tmp/TripalGFF3Import_BPDhlt
ERROR: The "Target" attribute is incorreclty formatted for the feature "<em class="placeholder"></em>."
 [error]  ERROR: The "Target" attribute is incorreclty formatted for the feature "<em class="placeholder"></em>." 
[site http://default] [TRIPAL] ERROR: The "Target" attribute is incorreclty formatted for the feature "<em class="placeholder"></em>."
```

The gff3 file looks like this:
```
seqid   source     type     start       end     score      strand     phase                   ID          Parent         Gap      Target        Name           Alias        _AED         _QI       _eAED
  <factor> <factor> <factor> <integer> <integer> <numeric> <character> <integer>          <character> <CharacterList> <character> <character> <character> <CharacterList> <character> <character> <character>
1     Chr4    maker     exon     85166     85390        NA           +        NA PG4T021925.V3:exon:0   PG4T021925.V3          NA          NA          NA                          NA          NA          NA
2     Chr4    maker     CDS      85166     85390        NA           +         0    PG4T021925.V3:cds   PG4T021925.V3          NA          NA          NA                          NA          NA          NA
3     Chr4    maker     exon     26182     26329        NA           +        NA PG4T021918.V3:exon:6   PG4T021918.V3          NA          NA          NA                          NA          NA          NA
4     Chr4    maker     exon     26842     27170        NA           +        NA PG4T021918.V3:exon:7   PG4T021918.V3          NA          NA          NA                          NA          NA          NA
5     Chr4    maker     exon     27329     27719        NA           +        NA PG4T021918.V3:exon:8   PG4T021918.V3          NA          NA          NA                          NA          NA          NA
6     Chr4    maker     exon     27809     27965        NA           +        NA PG4T021918.V3:exon:9   PG4T021918.V3          NA          NA          NA                          NA          NA          NA
```
I get that smth is wrong with the Target field of the file. But what?
Will be very grateful for any help with this...

Alex

**Comment by dsenalik** [comment]
What program generated that file? That file only partly matches gff3 format, and maybe there is a converter program for it. For instance column 1 should not be there, the ```NA``` values should be ```.```, and finally all the remaining columns should be combined in the last column probably something like
```ID=PG4T021925.V3%3Aexon%3A0;Parent=PG4T021925.V3```
The format is described here
https://useast.ensembl.org/info/website/upload/gff3.html

**Comment by alslonik** [comment]
Yes, sure, sorry, the actual text view of the file is this:
```
##gff-version 3
##source-version rtracklayer 1.50.0
##date 2021-05-25
Chr4	maker	exon	85166	85390	.	+	.	ID=PG4T021925.V3:exon:0;Parent=PG4T021925.V3;Gap=NA;Target=NA;
Chr4	maker	CDS	85166	85390	.	+	0	ID=PG4T021925.V3:cds;Parent=PG4T021925.V3;Gap=NA;Target=NA;
Chr4	maker	exon	26182	26329	.	+	.	ID=PG4T021918.V3:exon:6;Parent=PG4T021918.V3;Gap=NA;Target=NA;
Chr4	maker	exon	26842	27170	.	+	.	ID=PG4T021918.V3:exon:7;Parent=PG4T021918.V3;Gap=NA;Target=NA;
Chr4	maker	exon	27329	27719	.	+	.	ID=PG4T021918.V3:exon:8;Parent=PG4T021918.V3;Gap=NA;Target=NA;
Chr4	maker	exon	27809	27965	.	+	.	ID=PG4T021918.V3:exon:9;Parent=PG4T021918.V3;Gap=NA;Target=NA;
Chr4	maker	exon	28071	28129	.	+	.	ID=PG4T021918.V3:exon:10;Parent=PG4T021918.V3;Gap=NA;Target=NA;
```
The previous view is the view that I am using for filtering, it is R.

**Comment by laceysanderson** [comment]
Hi @alslonik, The target should include the uniquename of a feature in the database or the ID of a previous line in the GFF3 file and the start and end coordinates. If you do not have a target for a particular line then you should leave it out entirely.

For example, if your use of `NA` implies their is no value for that tag then your above snippet of GFF3 should look like this:

```
##gff-version 3
##source-version rtracklayer 1.50.0
##date 2021-05-25
Chr4	maker	exon	85166	85390	.	+	.	ID=PG4T021925.V3:exon:0;Parent=PG4T021925.V3
Chr4	maker	CDS	85166	85390	.	+	0	ID=PG4T021925.V3:cds;Parent=PG4T021925.V3
Chr4	maker	exon	26182	26329	.	+	.	ID=PG4T021918.V3:exon:6;Parent=PG4T021918.V3
Chr4	maker	exon	26842	27170	.	+	.	ID=PG4T021918.V3:exon:7;Parent=PG4T021918.V3
Chr4	maker	exon	27329	27719	.	+	.	ID=PG4T021918.V3:exon:8;Parent=PG4T021918.V3
Chr4	maker	exon	27809	27965	.	+	.	ID=PG4T021918.V3:exon:9;Parent=PG4T021918.V3
Chr4	maker	exon	28071	28129	.	+	.	ID=PG4T021918.V3:exon:10;Parent=PG4T021918.V3
```

Also note there is no need for the last semi-colon in the attributes column.

A correctly formatted target example would be:

```
ctg123 est EST_match 1050 1500 . + . ID=Match1;Name=agt830.5;Target=agt830.5 1 451
ctg123 est EST_match 3000 3202 . + . ID=Match1;Name=agt830.5;Target=agt830.5 452 654
ctg123 est EST_match 5410 5500 . - . ID=Match2;Name=agt830.3;Target=agt830.3 505 595
ctg123 est EST_match 7000 7503 . - . ID=Match2;Name=agt830.3;Target=agt830.3 1 504
ctg123 est EST_match 1050 1500 . + . ID=Match3;Name=agt221.5;Target=agt221.5 1 451
ctg123 est EST_match 5000 5500 . + . ID=Match3;Name=agt221.5;Target=agt221.5 452 952
ctg123 est EST_match 7000 7300 . + . ID=Match3;Name=agt221.5;Target=agt221.5 953 1253
```

You can read more about this here: http://gmod.org/wiki/GFF3#Alignments

**Comment by alslonik** [comment]
Thanks!  I managed to upload gff after correcting the Targets and the semicolon at the end of the lines. Now I want to publish a table of genes from my organism. How do i do it?  Also, Fasta uploader is not available yet for tripal 4, am I correct?

**Comment by dsenalik** [comment]
Correct, publish function is not yet implemented, and fasta uploader still needs to be implemented. Sorry 🙁

**Comment by spficklin** [comment]
@alslonik what are you thinking this table should look like?

**Comment by alslonik** [comment]
I think there should be an option to publish it as a table with the columns that one choses. I.e.  -  gene IDs only or with attributes as well (and than functional annotation can be placed there, and maybe more info) and the option to link each gene to Jbrowse. And an option to search for a gene with each one of the values published.

**Comment by spficklin** [comment]
@alslonik years ago Tripal did have such a feature for organism pages. It had a summary of genes and charts for functional enrichment. That worked fine until we started getting more than one genome assembly per organism or multiple transcriptome versions.  Also, the gene tables were so big (although we did have paging) and the feedback was they weren't very helpful because with tens of thousands of genes it was difficult to find anything.  It all got messy. 

For searching genes there is support for that.  Tripal v3 comes with a built-in Drupal Views search form that lets you search for genes. For Tripal v4 we will be integrating ElasticSearch and this is coming down the pipe.
